### PR TITLE
Configuration and priority

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
DNM, for discussion

### CODE REVIEW

File: src/gotcha.c [hunk 1]
```diff
@@ -204,3 +260,37 @@ GOTCHA_EXPORT enum gotcha_error_t gotcha_wrap(struct gotcha_binding_t* user_bind
...
+  else{
+    debug_printf(1, "Invalid property being configured on tool %s\n", tool_name);
+    return GOTCHA_INVALID_CONFIGURATION;
+  }
+  return GOTCHA_SUCCESS;
+}
+
+GOTCHA_EXPORT enum gotcha_error_t gotcha_set_priority(const char* tool_name, int value){
+  enum gotcha_error_t error_on_call = GOTCHA_SUCCESS;
+  enum gotcha_error_t error_on_set = gotcha_configure_int(tool_name, GOTCHA_PRIORITY, value);
```
Comment (DavidPoliakoff) [inline src/gotcha.c]: Make smrter: just return the error code, don't do this weird logic

File: src/tool.c [hunk 2]
```diff
@@ -102,3 +148,93 @@ binding_t *get_tool_bindings(tool_t *tool)
...
+  if(key==GOTCHA_PRIORITY){
+    int current_priority = config.priority;
+    if(current_priority!=UNSET_PRIORITY)
+    *((int*)(data)) = config.priority; 
+  }
+  return GOTCHA_SUCCESS;
+
+}
+
+enum gotcha_error_t get_configuration_value(const char* tool_name, enum gotcha_config_key_t key, void* data){
```
Comment (DavidPoliakoff) [inline src/tool.c]: data: thing that stores the return value

File: src/tool.c [hunk 3]
```diff
@@ -102,3 +148,93 @@ binding_t *get_tool_bindings(tool_t *tool)
...
+  struct gotcha_configuration_t config = get_default_configuration();
+  if(key==GOTCHA_PRIORITY){
+    int current_priority = config.priority;
+    if(current_priority!=UNSET_PRIORITY)
+    *((int*)(data)) = config.priority; 
+  }
+  return GOTCHA_SUCCESS;
+
+}
+
```
Comment (DavidPoliakoff) [inline src/tool.c]: Just add gotcha_get_priority which returns an int

File: src/tool.c [hunk 4]
```diff
@@ -102,3 +148,93 @@ binding_t *get_tool_bindings(tool_t *tool)
 {
    return tool->binding;
 }
+
+/**
+ * TODO DO-NOT-MERGE "/" should be a macro of possible separators
+ */
+struct gotcha_configuration_t get_configuration_for_tool(const char* tool_name_in){
```
Comment (DavidPoliakoff) [inline src/tool.c]: THOU SHALT FREE MEMORY

File: src/gotcha.c [hunk 5]
```diff
@@ -184,11 +239,12 @@ GOTCHA_EXPORT enum gotcha_error_t gotcha_wrap(struct gotcha_binding_t* user_bind
   }
 
   gotcha_prepare_symbols(bindings, num_actions);
-
+  int* write_table = (int*)malloc(sizeof(int)*num_actions);
```
Comment (DavidPoliakoff) [inline src/gotcha.c]: Free and possibly eliminate write_table

**Comment by DavidPoliakoff** [comment]
@daboehme Sure, that's an option, but somebody is going to do a CMake like

```
   get_my_c_flags_from_package(some_package)
   add_subdirectory(gotcha)
```

Some_package will define -Werror but not the C99 std flag, and stuff will break. It's goofy, but I'm choosing to fix it.

**Comment by daboehme** [comment]
I would assume that `set(CMAKE_C_STANDARD 99), set(CMAKE_C_STANDARD_REQUIRED)` in Gotcha's CMakeLists.txt would fix that, and apply it to the current subdirectory only without propagating elsewhere.

**Comment by DavidPoliakoff** [comment]
Let me look into that, I think you're correct. Appreciate the advice

**Comment by DavidPoliakoff** [comment]
https://cmake.org/cmake/help/v3.0/search.html?q=cmake_c_standard&check_keywords=yes&area=default

It appears that option was introduced in CMake 3.1, I know we still have machines with a default of 2.8. This is low enough level that I'm willing to eat the annoyance of predeclaring iterator variables.

**Comment by daboehme** [comment]
😞

**Comment by DavidPoliakoff** [comment]
@mplegendre this one is probably ready for more conversation tomorrow. 

We now have the opaque handle in our user's gotcha_binding_t's, a hash table from function name to chain of internal binding chains, and a system that respects/uses them

**Pull request merged**